gateway/ln-gateway/src/lnd.rs

```diff
@@ -0,0 +1,302 @@
+                                // outcomes on `complete_htlc` rpc
+                                outcomes.lock().await.insert(
+                                    intercepted_htlc_id,
+                                    (Arc::new(lnd_tx.clone()), htlc.incoming_circuit_key),
+                                );
+
+                                None
+                            }
+                            Err(e) => {
+                                error!("Failed to send HTLC to gatewayd for processing: {:?}", e);
+                                Some(cancel_intercepted_htlc(htlc.incoming_circuit_key))
+                            }
+                        }
+                    };
+
+                if response.is_some() {
+                    // TODO: Consider retrying this if the send fails
+                    let _ = lnd_tx.send(response.unwrap()).await.map_err(|e| {
+                        error!("Failed to send response to LND: {:?}", e);
+                        // The HTLC will timeout and LND will automatically cancel it.
+                        GatewayError::Other(anyhow!("Failed to send response to LND"))
+                    });
+                }
+            }
+
+            info!("HTLC subscription stream ended");
+        })
+        .await;
+
+        Ok(Box::pin(ReceiverStream::new(gwd_rx)))
+    }
+
+    async fn complete_htlc(
+        &self,
+        outcome: CompleteHtlcsRequest,
+    ) -> crate::Result<CompleteHtlcsResponse> {
+        let CompleteHtlcsRequest {
+            action,
+            intercepted_htlc_id,
+        } = outcome;
+
+        let hash = match sha256::Hash::from_slice(&intercepted_htlc_id) {
+            Ok(hash) => hash,
+            Err(e) => {
+                error!("Invalid intercepted_htlc_id: {:?}", e);
+                return Err(GatewayError::LnRpcError(tonic::Status::invalid_argument(
+                    e.to_string(),
+                )));
+            }
+        };
+
+        info!("Completing htlc with reference, {:?}", hash);
+
+        if let Some((outcome, incoming_circuit_key)) = self.outcomes.lock().await.remove(&hash) {
+            let htlca_res = match action {
+                Some(Action::Settle(Settle { preimage })) => ForwardHtlcInterceptResponse {
+                    incoming_circuit_key,
+                    action: 0,
+                    preimage,
+                    failure_message: vec![],
+                    failure_code: 0,
+                },
+                Some(Action::Cancel(Cancel { reason: _ })) => cancel_intercepted_htlc(None),
+                None => {
+                    error!("No action specified for intercepted htlc id: {:?}", hash);
+                    return Err(GatewayError::LnRpcError(tonic::Status::internal(
+                        "No action specified on this intercepted htlc",
+                    )));
+                }
+            };
+
+            // TODO: Consider retrying this if the send fails
```

Comment:
Nit: this looks like the same code as above, maybe we could make it a function

Reply:
[#1936](https://github.com/fedimint/fedimint/issues/1936) is the preferred fix